### PR TITLE
[ENH] Add `reverse` parameter to `cumsum()`, `cumprod()`, `cummin()` and `cummax()`

### LINKED ISSUE #3279: Implement cumulative functions
The list of functions to be implemented and the corresponding PR's
- [x] `cumsum()` https://github.com/h2oai/datatable/pull/3257
- [x] `cumprod()` https://github.com/h2oai/datatable/pull/3304
- [x] `cummax()` https://github.com/h2oai/datatable/pull/3288
- [x] `cummin()` https://github.com/h2oai/datatable/pull/3288
- [x] `cumcount()` https://github.com/h2oai/datatable/pull/3310
- [x] `ngroup()` - not strictly cumulative https://github.com/h2oai/datatable/pull/3310
- [x] `fillna()` for forward/backward fill https://github.com/h2oai/datatable/pull/3311
- [x] `fillna()` for filling with a value https://github.com/h2oai/datatable/pull/3344 
~~- [ ] rank~~ continued on #3148 
~~- [ ] rolling aggregations~~    continued on #1500

### PR DESCRIPTION
**Pull request by samukweku**
Add `reverse` parameter to control direction of cumulative function's calculations: 
- when `False`, calculation is done from top to bottom (default);
- when `True`, calculation is done from bottom to top.

Сloses #3279

### CODE REVIEW

File: docs/api/dt/cummax.rst [hunk 1]
```diff
@@ -15,6 +15,11 @@
     cols: FExpr
         Input data for cumulative maximum calculation.
 
+    reverse: bool
+        If ``False``, the cumulative maximum is computed from the last row
+        to the first row. if ``True``, the cumulative maximum is computed 
+        from the first row to the last row.
```
Comment (oleksiyskononenko) [inline docs/api/dt/cummax.rst:21]: I guess it is vice versa: 
- `False` -> from the first to the last row;
- `True` -> from the last to the first row.

Also, `reverse` parameter is missing in the function signature.

File: docs/api/dt/cummin.rst [hunk 2]
```diff
@@ -15,6 +15,11 @@
     cols: FExpr
         Input data for cumulative minimum calculation.
 
+    reverse: bool
+        If ``False``, the cumulative minimum is computed from the last row
+        to the first row. if ``True``, the cumulative minimum is computed 
+        from the first row to the last row.
```
Comment (oleksiyskononenko) [inline docs/api/dt/cummin.rst:21]: Same here.

File: docs/api/dt/cumprod.rst [hunk 3]
```diff
@@ -16,6 +16,11 @@
     cols: FExpr
         Input data for cumulative product calculation.
 
+    reverse: bool
+        If ``False``, the cumulative product is computed from the last row
+        to the first row. if ``True``, the cumulative product is computed 
+        from the first row to the last row.
```
Comment (oleksiyskononenko) [inline docs/api/dt/cumprod.rst:22]: Same here.

File: docs/api/dt/cumsum.rst [hunk 4]
```diff
@@ -16,6 +16,11 @@
     cols: FExpr
         Input data for cumulative sum calculation.
 
+    reverse: bool
+        If ``False``, the cumulative sum is computed from the last row
+        to the first row. if ``True``, the cumulative sum is computed 
+        from the first row to the last row.
```
Comment (oleksiyskononenko) [inline docs/api/dt/cumsum.rst:22]: Same here.

File: src/core/column/cumminmax.h [hunk 5]
```diff
@@ -57,24 +57,42 @@ class CumMinMax_ColumnImpl : public Virtual_ColumnImpl {
...
               } else {
-                data[i] = (res_valid && data[i - 1] > val)? data[i - 1] : val;
+                data[i] = (res_valid && data[i + 1] > val)? data[i + 1] : val;
               }
               res_valid = true;
             } else {
-              data[i] = data[i - 1];
+              data[i] = data[i + 1];
+            }
             }
```
Comment (oleksiyskononenko) [inline src/core/column/cumminmax.h:76]: Issues with the indentation.

File: src/core/column/cumminmax.h [hunk 6]
```diff
@@ -57,24 +57,42 @@ class CumMinMax_ColumnImpl : public Virtual_ColumnImpl {
...
+                if (MIN) {
+                  data[i] = (res_valid && data[i - 1] < val)? data[i - 1] : val;
+                } else {
+                  data[i] = (res_valid && data[i - 1] > val)? data[i - 1] : val;
+                }
+                res_valid = true;
+              } else {
+                data[i] = data[i - 1];
+              }
+            }}
```
Comment (oleksiyskononenko) [inline src/core/column/cumminmax.h:94]: Same here. We normally align the opening and closing brackets.

File: src/core/column/cumsumprod.h [hunk 7]
```diff
@@ -56,6 +56,24 @@ namespace dt {
...
+            for (size_t i = i2-1; i-- > i1;) {
+              bool is_valid = col_.get_element(i, &val);
+              if (SUM) {
+                data[i] = data[i + 1] + (is_valid? val : 0);
+              } else {
+                data[i] = data[i + 1] * (is_valid? val : 1);
+              }
+              
+              }
+          } else {
```
Comment (oleksiyskononenko) [inline src/core/column/cumsumprod.h:76]: Brackets alignment issues.

File: tests/dt/test-cumminmax.py [hunk 8]
```diff
@@ -180,3 +182,21 @@ def test_cumminmax_groupby_complex():
...
+def test_cumminmax_groupby_reverse():
+    DT = dt.Frame([[3, 14, 15, 92, 6], ["a", "cat", "a", "dog", "cat"]])
+    DT_mm = DT[:, [cummin(f[0], reverse=True), cummax(f[0], True)], by(f[1])]
+
+    DT_ref = dt.Frame(
+                 {"C1" : ["a", "a", "cat", "cat", "dog"],
+                 "C0" : [3, 15, 6, 6, 92],
+                 "C2" : [15, 15, 14, 6, 92]}
+             )
+    assert_equals(DT_mm, DT_ref)
```
Comment (oleksiyskononenko) [inline tests/dt/test-cumminmax.py:202]: Newline is missing.

**Comment by oleksiyskononenko** [comment]
I guess I've fixed most of the issues mentioned in the comments. Please, take a look.

**Comment by samukweku** [comment]
thanks @oleksiyskononenko looks good

**Comment by samukweku** [comment]
@oleksiyskononenko I cant figure out the fail on test-dt.py line 1084, or its purpose.

File: src/core/column/cumminmax.h (hunk 6 above)
Comment (oleksiyskononenko) [reply src/core/column/cumminmax.h]: I'm using Sublime and 2-space indentation for C++ / 4-space indentation for Python. Then, once I press the tab button I get the required indentation depending on the language. As for the brackets, Sublime automatically indents the closing bracket (using the appropriate number of spaces) once you type in the opening one.

**Comment by oleksiyskononenko** [comment]
@samukweku Uh, don't worry about it. It is not relevant for this PR. This test fails from time to time due to the system performance. We probably need to adjust it at some point.

**Review by oleksiyskononenko — APPROVED** [review]

**Pull request merged**